src/libawkward/array/RegularArray.cpp

```diff
@@ -515,6 +515,111 @@ namespace awkward {
     throw std::invalid_argument("slice items can have all fixed-size dimensions (to follow NumPy's slice rules) or they can have all var-sized dimensions (for jagged indexing), but not both in the same slice item");
   }
 
+  const std::shared_ptr<Content> RegularArray::rpad(int64_t length, int64_t axis) const {
+    int64_t toaxis = axis_wrap_if_negative(axis);
+    std::shared_ptr<Content> out = content();
+    // If the content is not an integer multiple of size,
+    // then the length of the RegularArray is truncated
+    // to the largest integer multiple
+    int64_t n = std::floor(out.get()->length()/size());
+    int64_t fromlength = n*size();
+
+    if (toaxis == 0) {
+      if (length > size()) {
+        int64_t tolength = (length - size())*size();
+
+        if (out.get()->isindexed()) {
+          out = out.get()->getitem_range_nowrap(0, fromlength);
+          out = out.get()->rpad(tolength, toaxis);
+        }
+        else {
+          (length >= out.get()->length()/size()) ? tolength = length*size() : tolength = out.get()->length() - length*size();
+          Index64 outindex(tolength);
+          struct Error err = awkward_index_rpad(
+            outindex.ptr().get(),
+            fromlength,
+            tolength);
+          util::handle_error(err, classname(), identities_.get());
+
+          out = std::make_shared<IndexedOptionArray64>(identities_, parameters_, outindex, out);
+        }
+      }
+      else {  // clip it
+        int64_t tolength = length*size();
+        out = out.get()->getitem_range_nowrap(0, tolength);
+      }
+
+      return std::make_shared<RegularArray>(identities_, parameters_, out, size());
+    }
+    else if (toaxis == 1) {
+      if(out.get()->isindexed()) {
```

Comment:
This doesn't actually benefit from `isindexed` (except in that non-indexed types get to skip five `dynamic_cast` checks; a bit like a tree-lookup of two levels, but we're not concerned about that kind of performance here). Even a virtual IndexedArrayBase wouldn't help here because each section still needs to choose different functions, based on specialization.

It could be simplified by retemplatizing `awkward_indexedarray_inject_rpad_from32` etc. as `util::awkward_indexedarray_inject_rpad_from<T>`. You'll see that I've done that for a lot of these functions in src/libawkward/util.cpp. It's a bit ridiculous: an originally templated function is untemplatized to pass through a C interface and then retemplatized to simplify the C++ code, but it's the best I can think of while maintaining the requirement of having a C interface to the kernels.

Reply:
Oh, I see that you are, in fact, retemplatizing it. I didn't realize it because you didn't make the template parameter explicit here—it's being inferred from the argument types. Conventionally, I've made the template parameter explicit as a form of documentation, but I suppose that's not necessary.

Reply:
removed as not needed